drivers/interrupt_controller/intc_scr_ipic.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2023 Syntacore
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT scr_ipic
+
+/**
+ * @file
+ * @brief Syntacore Integrated Programmable Interrupt Controller (IPIC) Interface
+ *        for RISC-V processors
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/init.h>
+#include <zephyr/arch/riscv/csr.h>
+#include <zephyr/device.h>
+
+#include <zephyr/sw_isr_table.h>
+#include <zephyr/drivers/interrupt_controller/intc_scr_ipic.h>
+#include <zephyr/irq.h>
+
+#define IPIC_IRQ_LEVEL_HIGH	0
+#define IPIC_IRQ_CLEAR_PENDING	(BIT(0))
+#define IPIC_IRQ_ENABLE	(BIT(1))
+#define IPIC_IRQ_PRIV_MMODE	(BIT(9) | BIT(8))
+#define IPIC_IRQ_LN_OFFS	12
+
+#define PLF_IPIC_MBASE	0xbf0
+#define IPIC_CISV	(PLF_IPIC_MBASE + 0)
+#define IPIC_CICSR	(PLF_IPIC_MBASE + 1)
+#define IPIC_IPR	(PLF_IPIC_MBASE + 2)
+#define IPIC_ISVR	(PLF_IPIC_MBASE + 3)
+#define IPIC_EOI	(PLF_IPIC_MBASE + 4)
+#define IPIC_SOI	(PLF_IPIC_MBASE + 5)
+#define IPIC_IDX	(PLF_IPIC_MBASE + 6)
+#define IPIC_ICSR	(PLF_IPIC_MBASE + 7)
+#define IPIC_VOID_VEC	16
+
+#define MK_IRQ_CFG(line, mode, flags) ((mode) | (flags) | ((line) << IPIC_IRQ_LN_OFFS))
+
+static unsigned long ipic_irq_current_vector(void)
+{
+	return csr_read(IPIC_CISV);
+}
+
+static int ipic_irq_setup(int irq_vec, int line, int mode, int flags)
+{
+	if (CONFIG_IPIC_STATIC_LINE_MAPPING || irq_vec < 0) {
+		irq_vec = line;
+	}
+
+	csr_write(IPIC_IDX, irq_vec);
+	csr_write(IPIC_ICSR, MK_IRQ_CFG(line, mode, flags | IPIC_IRQ_CLEAR_PENDING));
+
+	return irq_vec;
+}
+
+static void ipic_irq_reset(int irq_vec)
+{
+	ipic_irq_setup(irq_vec, CONFIG_EXT_IPIC_IRQ_LN_NUM,
+			IPIC_IRQ_PRIV_MMODE, IPIC_IRQ_CLEAR_PENDING);
+}
+
+static void ipic_irq_enable(unsigned int irq_vec)
+{
+	csr_write(IPIC_IDX, irq_vec);
+	const unsigned long state =
```

Comment:
please (here and everywhere) put variables declaration at the entry of the function (on top).